driver-core/src/main/java/com/datastax/driver/core/policies/HostFilterPolicy.java

```diff
@@ -0,0 +1,226 @@
+
+        childPolicy.init(cluster, whiteHosts);
+    }
+
+    /**
+     * Return the HostDistance for the provided host.
+     *
+     * @param host the host of which to return the distance of.
+     * @return {@link HostDistance#IGNORED} if {@code host} is not matching the predicate, the HostDistance
+     * as returned by the wrapped policy otherwise.
+     */
+    @Override
+    public HostDistance distance(Host host) {
+        return predicate.apply(host)
+             ? childPolicy.distance(host)
+             : HostDistance.IGNORED;
+    }
+
+    /**
+     * Returns the hosts to use for a new query.
+     * <p>
+     * It is guaranteed that only hosts matching the predicate will be returned.
+     *
+     * @param loggedKeyspace the currently logged keyspace (the one set through either
+     * {@link Cluster#connect(String)} or by manually doing a {@code USE} query) for
+     * the session on which this plan need to be built. This can be {@code null} if
+     * the corresponding session has no keyspace logged in.
+     * @param statement the query for which to build a plan.
+     */
+    @Override
+    public Iterator<Host> newQueryPlan(String loggedKeyspace, Statement statement) {
+        // Just delegate to the child policy, since we filter the hosts not white
+        // listed upfront, the child policy will never see a host that is not white
+        // listed and thus can't return one.
+        return childPolicy.newQueryPlan(loggedKeyspace, statement);
+    }
+
+    @Override
+    public void onUp(Host host) {
+        if (predicate.apply(host))
+            childPolicy.onUp(host);
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public void onSuspected(Host host) {
+        if (predicate.apply(host))
+            childPolicy.onSuspected(host);
+    }
+
+    @Override
+    public void onDown(Host host) {
+        if (predicate.apply(host))
+            childPolicy.onDown(host);
+    }
+
+    @Override
+    public void onAdd(Host host) {
+        if (predicate.apply(host))
+            childPolicy.onAdd(host);
+    }
+
+    @Override
+    public void onRemove(Host host) {
+        if (predicate.apply(host))
+            childPolicy.onRemove(host);
+    }
+
+    @Override
+    public void close() {
+        if (childPolicy instanceof CloseableLoadBalancingPolicy)
+            ((CloseableLoadBalancingPolicy)childPolicy).close();
+    }
+    
+    /**
+     * Create a new policy that wraps the provided child policy but only "allow/forbids" hosts
+     * from the provided list.
+     *
+     * @param childPolicy the wrapped policy.
+     * @param hosts the hosts.
+     * @param allow when true allow hosts from the provided list to be connected to 
+     * (whether they will get connected to or not depends on the child policy), otherwise forbids these hosts
+     */
+    public static HostFilterPolicy mkHostAddressListPolicy( LoadBalancingPolicy childPolicy, Collection<InetSocketAddress> hosts, boolean allow) {
```

Comment:
It would be a bit more readable to split into two methods in order to avoid the boolean argument. For example `fromHostWhiteList` and `fromHostBlackList` (same thing for DC methods).